x/ante/check_refund.go

```diff
@@ -35,62 +36,69 @@ func NewCheckRefundFeeDecorator(registry cdctypes.InterfaceRegistry, ak antetype
 func (d CheckRefundFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
 	msgs := tx.GetMsgs()
 
-	// reject unregistered refundable msg
-	for _, msg := range msgs {
-		switch msg := msg.(type) {
-		case *rewardtypes.RefundMsgRequest:
-			if !msgRegistered(d.registry, msg.InnerMessage.GetTypeUrl()) {
-				return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, fmt.Sprintf("%T is not refundable", msg))
-			}
-		}
+	if !anyRefundable(msgs) {
+		return next(ctx, tx, simulate)
 	}
 
-	if d.qualifyForRefund(ctx, msgs) {
-		feeTx, ok := tx.(sdk.FeeTx)
-		if !ok {
-			return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
-		}
+	if err := d.qualifyForRefund(ctx, msgs); err != nil {
+		return ctx, err
+	}
 
-		fee := feeTx.GetFee()
-		if len(fee) > 0 {
-			req := msgs[0].(*rewardtypes.RefundMsgRequest)
-			err := d.reward.SetPendingRefund(ctx, *req, fee[0])
-			if err != nil {
-				return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, err.Error())
-			}
-		}
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "tx must be a FeeTx")
+	}
 
+	fee := feeTx.GetFee()
+	if len(fee) > 0 {
+		req := msgs[0].(*rewardtypes.RefundMsgRequest)
+		err := d.reward.SetPendingRefund(ctx, *req, fee[0])
+		if err != nil {
+			return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, err.Error())
+		}
 	}
 
 	return next(ctx, tx, simulate)
 }
 
-func (d CheckRefundFeeDecorator) qualifyForRefund(ctx sdk.Context, msgs []sdk.Msg) bool {
-	if len(msgs) != 1 {
+func anyRefundable(msgs []sdk.Msg) bool {
+	if len(msgs) == 0 {
 		return false
 	}
 
-	switch msg := msgs[0].(type) {
-	case *rewardtypes.RefundMsgRequest:
-		if !msgRegistered(d.registry, msg.InnerMessage.GetTypeUrl()) {
-			return false
+	for _, msg := range msgs {
+		switch msg.(type) {
+		case *rewardtypes.RefundMsgRequest:
+			return true
 		}
+	}
+	return false
+}
 
-		// Validator must be bonded
-		sender := msg.GetSigners()[0]
-		validatorAddr := d.snapshotter.GetOperator(ctx, sender)
-		if validatorAddr == nil {
-			return false
-		}
-		validator := d.staking.Validator(ctx, validatorAddr)
-		if validator == nil || !validator.IsBonded() {
-			return false
-		}
+func (d CheckRefundFeeDecorator) qualifyForRefund(ctx sdk.Context, msgs []sdk.Msg) error {
+	for _, msg := range msgs {
+		switch msg := msg.(type) {
+		case *rewardtypes.RefundMsgRequest:
+			if !msgRegistered(d.registry, msg.InnerMessage.TypeUrl) {
+				return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, fmt.Sprintf("message type %s is not refundable", msg.InnerMessage.TypeUrl))
+			}
+
+			sender := msg.GetSigners()[0]
```

Comment:
From what I understand so far, a `Tx` can have multiple signers. For each `Msg`, Cosmos verifies that all signers listed in `GetSigners` have signed the `Tx`, but the converse isn't necessarily true. Specifically, can we have the scenario where the `Tx` is signed by `acc1` and `acc2`, but it contains `msg1` signed by `acc1` and `msg2` signed by `acc2`? If true, not sure if there are any concerning implications, but it seems like a weird scenario, in which case might make sense to verify that `Tx` has a single signer or that all `Msg` senders are the same.

Reply:
it is possible, but that just means that the first signer gets the reward. The way I see it is this: having multiple signers doesn't break anything, so it is not security relevant. We expect refund txs to be signed by a single signer and the code is set up that way. But if someone wants to go out of their way to create  a weird tx with multiple signers, that's on them

Reply:
Yeah, doesn't seem exploitable.